### ISSUE TITLE
Dialog API Error with OfficeJS Outlook Add-in on Outlook 2016 for Windows

### ISSUE THREAD

**Issue by TheSamsterZA** [OfficeDev/office-js#633]
## Expected Behavior
Creating, attaching event handlers, opening, messaging between parent and child, and closing a dialog should work.

## Current Behavior

On Outlook 2016 for Windows, on the open dialog, I am seeing the following error in the console:
```
Error: AgaveSink could not find a suitable sink to use
```

## Steps to Reproduce, or Live Example

* Link to live example: https://github.com/TheSamsterZA/office-js-dialog-api-bug

* Additional details:
  1. `npm install`
  2. `npm start`
  3. Sideload the provided manifest
  4. On Outlook 2016 for Windows, open the add-in task pane, and attach the debugger (1)
  5. Open the dialog, and attach another debugger (1) to that
  5. Click the button on the dialog: the dialog should close; instead it remains open
  6. Notice the error 

(1) Use the [Microsoft Edge DevTools](https://www.microsoft.com/en-au/p/microsoft-edge-devtools-preview/9mzbfrmz0mnj?activetab=pivot:overviewtab) to attach debuggers.

## Context

  In my actual add-in that is currently live on AppSource, I am opening a dialog to a third-party authentication provider. After a successful login on the provider, I am redirected back to my add-in within the dialog. I [Office.context.ui.messageParent](https://docs.microsoft.com/en-us/javascript/api/office/office.ui?view=office-js#messageparent-message-) but nothing happens and the error above is thrown. The dialog remains open, and I do not receive the token from the dialog. The login therefore fails and the user cannot use my add-in.

## Your Environment

* Platform: PC desktop
* Host: Outlook
* Office version number: Outlook 2016 for Windows Version 1908 (Build 11911.20000 Click-to-Run), Office Insider
* Operating System: Windows 10 Version 1903 (OS Build 18362.239)
* Browser (if using Office Online): N/A

**NOTE:** [Since my environment supports it](https://docs.microsoft.com/en-au/office/dev/add-ins/concepts/browsers-used-by-office-web-add-ins), Outlook is using the `Microsoft Edge WebView, browser engine EdgeHTML 18.18362 (not Chromium)`.

## Useful logs

- [x] Console errors
```
Error: AgaveSink could not find a suitable sink to use
 at e.prototype.failToInitialize (https://appsforoffice.microsoft.com/lib/1.1/hosted/telemetry/oteljs_agave.js:1:71169)
 at e.prototype.initialize (https://appsforoffice.microsoft.com/lib/1.1/hosted/telemetry/oteljs_agave.js:1:70452)
 at e (https://appsforoffice.microsoft.com/lib/1.1/hosted/telemetry/oteljs_agave.js:1:70227)
 at e.createInstance (https://appsforoffice.microsoft.com/lib/1.1/hosted/telemetry/oteljs_agave.js:1:70361)
 at c.create (https://appsforoffice.microsoft.com/lib/1.1/hosted/office.js:18:16634)
 at e (https://appsforoffice.microsoft.com/lib/1.1/hosted/office.js:18:17056)
 at Anonymous function (https://appsforoffice.microsoft.com/lib/1.1/hosted/office.js:18:17204)
 at t.prototype.invoke (https://outlookapp.test.titlex.com.au/polyfills.77e5fe1dedbbcaec49d6.bundle.js:1:88956)
 at e.prototype.run (https://outlookapp.test.titlex.com.au/polyfills.77e5fe1dedbbcaec49d6.bundle.js:1:84204)
 at Anonymous function (https://outlookapp.test.titlex.com.au/polyfills.77e5fe1dedbbcaec49d6.bundle.js:1:95475)
```

I spent some time trying to debug this, and noticed in `https://appsforoffice.microsoft.com/lib/1.1/hosted/telemetry/oteljs_agave.debug.js`:

```
3806    AgaveSink.prototype.initialize = function() {
3807        if (!this.isTelemetryEnabled()) {
3808            this.failToInitialize(E_TELEMETRY_DISABLED);
3809        }
3810        if (OutlookSink_OutlookSink.isSupported()) {
3811            this.connectOutlookSink();
3812        } else if (SdxWacSink_SdxWacSink.isSupported()) {
3813            this.connectSdxWacSink();
3814        } else if (isWacAgave() || typeof OfficeExtension === "undefined") {
3815            if (canSendToAria()) {
3816                this.connectAriaSink();
3817            } else {
3818                this.failToInitialize();
3819            }
3820        } else {
3821            this._awaitingInitialization = true;
3822            getRichApiSink(false, this.onGetRichApi.bind(this));
3823        }
3824    };
```

I believe the bug is in here or in one of the functions called within here. There are assumptions about retrieving the host information which I believe cannot be done when in an OfficeJS dialog.

**Comment by niteshgolchha85** [comment]
Looks like the same issue as reported here: https://github.com/OfficeDev/office-js/issues/612

**Comment by niteshgolchha85** [comment]
Can you confirm if you have the latest office.js version?

**Comment by TheSamsterZA** [comment]
Hello @niteshgolchha85,

I don't think #612 is the exact same issue, or if it is, then it still is not resolved. I am still seeing the error I mentioned above with the stacktrace.

I do not believe I have [KB4464534](https://support.microsoft.com/en-us/help/4464534/security-update-for-office-2016-july-9-2019) installed, either.

I am pretty sure I'm on the latest released version of office.js, my add-in simply requests it from `https://appsforoffice.microsoft.com/lib/1.1/hosted/office.js`. which then goes on to request:

`https://appsforoffice.microsoft.com/lib/1.1/hosted/outlook-win32-16.02.js`
`https://appsforoffice.microsoft.com/lib/1.1/hosted/en-us/office_strings.js`
`https://appsforoffice.microsoft.com/lib/1.1/hosted/telemetry/oteljs.js`
`https://appsforoffice.microsoft.com/lib/1.1/hosted/en-us/outlook_strings.js`
`https://appsforoffice.microsoft.com/lib/1.1/hosted/telemetry/oteljs_agave.js`

among others.

Please let me know if I can help in any way to get this resolved.

**Comment by niteshgolchha85** [comment]
#612 has been resolved now. Can you try now?

**Comment by jargil** [comment]
@TheSamsterZA - I cannot reproduce this problem (see screenshot below, left debugger is the dialog, right debugger is the add-in main window), I see both displayDialogAsync and messageParent APIs working just fine with the latest builds of Office. Could you try after making sure you have updated to the latest Office JS as well as you should be able to update to Office 11916.

![image](https://user-images.githubusercontent.com/10125816/61911024-0f5fe080-aeeb-11e9-95db-6d3e5f9790ff.png)

**Comment by TheSamsterZA** [comment]
Hello @jargil,

I have updated to Version 1908 (Build 11916.20000 Click-to-Run) but unfortunately, I still see the issue.

~~I noticed in your screenshot that you are using IE11 - perhaps update Windows to [Windows 10 1903 so that the Edge Webview is used instead](https://docs.microsoft.com/en-au/office/dev/add-ins/concepts/browsers-used-by-office-web-add-ins)?~~
My bad, I assumed that was IE11, but it could be Edge! :)

Please keep in mind that your tests must:

1. show the dialog with a different URL (my sample does this, pointing to `dialog.html` on the same domain)
2. the dialog must redirect to the main taskpane's URL (so both the dialog and taskpane have the same URL)
3. the dialog would then [messageParent](https://docs.microsoft.com/en-us/javascript/api/office/office.ui?view=office-js#messageparent-message-)
4. the taskpane would receive the message, process it, and finally [ask the dialog to close](https://docs.microsoft.com/en-us/javascript/api/office/office.dialog?view=office-js#close--)

Please let me know if you need further info.

**Comment by eladco898** [comment]
I have the same issue. It appeared in Windows Outlook Desktop client only after installing the latest Windows security update KB4464534.

**Comment by jargil** [comment]
@TheSamsterZA - thanks for the detailed repro steps. I still cannot reproduce. I followed your steps and then added multiple additional navigations to different urls within the same domain and I was able to call messageParent just fine. Could you double check your sample code? Maybe if you could post a sample repro using ScriptLab?

@eladco898 - there was a regression with our security update that was blocking certain APIs. This has been resolved, please try updating to the latest Office build and also make sure that you have the latest Office.js from our CDN.

**Comment by TheSamsterZA** [comment]
@jargil [Script Lab is not available in Outlook](https://appsource.microsoft.com/en-us/product/office/WA104380862).

Here is my updated list of versions:

> Platform: PC Desktop
> Host: Outlook
> Office version number: Outlook 2016 for Windows Version 1909 (Build 12001.20000 Click-to-Run), Office Insider
> Operating System: Windows 10 Version 1903 (OS Build 18362.239)

**NOTE:** [Since my environment supports it](https://docs.microsoft.com/en-au/office/dev/add-ins/concepts/browsers-used-by-office-web-add-ins), Outlook is using the Microsoft Edge WebView, browser engine EdgeHTML 18.18362 (not Chromium).

I am still seeing the error and stacktrace I mentioned above.

---
Below is example code I am [using in my sample on GitHub](https://github.com/TheSamsterZA/office-js-dialog-api-bug), that I have placed in Script Lab, running in **Word**. This works great:

```javascript
$("#run").click(() => tryCatch(run));

async function run() {

  const isInDialog: boolean = Office && Office.context && Office.context.document == null || false;

  if (isInDialog) {
    console.log(`In dialog - sending message to parent`);
    Office.context.ui.messageParent(JSON.stringify({ type: 'MessageFromDialog', data: 'Hello from dialog' }));
  }
  else {
  Office.context.ui.displayDialogAsync(
    window.location.href,
    { height: 50, width: 50 },
    (result) => {
      if (result.status === Office.AsyncResultStatus.Succeeded) {
        const dialog = result.value;
        dialog.addEventHandler(Office.EventType.DialogMessageReceived, (msg) => {
          console.log(`Office.EventType.DialogMessageReceived handler!`);
          if (msg) {
            console.log(msg);
            console.log("MESSAGE:", JSON.parse(msg.message));
            console.log(`Message received - closing dialog from parent`);
            dialog.close();
          }
        });

        dialog.addEventHandler(Office.EventType.DialogEventReceived, (evt) => {
          console.log(`Office.EventType.DialogEventReceived handler!`);
          if (evt) {
            console.log(evt);
          }
        });
      } else {
        // Oh no
        console.log(`ERROR SHOWING DIALOG`);
        console.log(result.error);
        console.log(result.diagnostics);
      }
    }
  );
  }
}

/** Default helper for invoking an action and handling errors. */
async function tryCatch(callback) {
  try {
    await callback();
  } catch (error) {
    // Note: In a production add-in, you'd want to notify the user through your add-in's UI.
    console.error(error);
  }
}
```

Hope this helps.

**Comment by YannickRe** [comment]
I'm encountering the same issue in Outlook included with Office 365 ProPlus. It started occurring when we were switched from IE WebView to Edge WebView.
My add-in still works in browser Edge, but fails in Edge WebView with the error `Error: AgaveSink could not find a suitable sink to use`.

As a result, my access token isn't passed to the add-in task pane and we are stuck.

**Comment by jargil** [comment]
@TheSamsterZA - The problem with your GitHub sample is that you have a couple of invalid characters in your dialog.html. After making below change, I see that upon clicking on "Send message to parent", the page navigates to Index.html, sends the message and closes:

diff --git a/src/dialog.html b/src/dialog.html
--- a/src/dialog.html
+++ b/src/dialog.html
@@ -21,7 +21,7 @@
     function onClick() {
       var host = window.location.host;
       var protocol = window.location.protocol;
\-      var url = \`${protocol}//${host}/index.html\`;
\+      var url = '${protocol}//${host}/index.html';
       console.log('Redirect back');
       window.location.assign(url);
     }

Please confirm so that I can go ahead and close this issue.

**Comment by YannickRe** [comment]
@jargil I would be VERY surprised that changing the backticks to normal single quotes would solve this issue, and even if it does then I believe it is a sign of a deeper rooted problem.
The backticks are a sign of template literals (https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals) that would allow for easy string interpolation. Now, if you change it to normal single quotes, the variable replacement will not happen anymore...

**Comment by jargil** [comment]
On a second inspection, the sample as is is throwing "SCRIPT1014: Invalid character dialog.html (24,17)" on the IE runtime (pre Windows May update) which is throwing your DOM into an error state (e.g. onClick is left undefined) and additionally blocking the dialog WebView instance from loading all the required JavaScript, thus the AgaveSink error you see. Without the string replacement happening this sample ends up navigating to https://localhost:3000/$%7Bprotocol%7D//$%7Bhost%7D/index.html which is simply index.html and works on the IE runtime (proper string replacement still needed).

@TheSamsterZA - when I run your sample in the Edge WebView it works as is and I don't see any AgaveSink error, messaging parent works as expected, and the dialog closes properly.

**Comment by TheSamsterZA** [comment]
@jargil Thanks for your detective work, that is an interesting find.

It is quite late here (11 PM) so I will verify your suggested fix tomorrow. Please hold off on closing this issue until then.

**Comment by jargil** [comment]
Note that I was incorrectly using the previous IE runtime when I posted the first part about changing the quotes, sorry about that. On Edge WebView I cannot reproduce what you're seeing so I'm not able to really help much other than confirm that your sample works for me.

**Comment by YannickRe** [comment]
@jargil That explains it, since IE doesn't support template literals.
I don't have an issue with the sample @TheSamsterZA posted, but I do have this issue in my add-in (only in the Edge WebView, not in Edge browser or any other browser).
I'll open a new issue once I'm able to provide a simplified, reproducible sample. I'm thinking mine has to do with a redirect to a different domain (for authentication), but I'll investigate further on my own first.

**Comment by jargil** [comment]
@TheSamsterZA - I cannot repro neither by following up your repro steps or by using your GitHub sample. Please let me know if you have any further questions or follow ups.

**Comment by TheSamsterZA** [comment]
@jargil I think it is premature to close this issue. I am seeing this on our live add-ins that are on AppSource. Here are replication steps:

1. Install [LEAP Mail Manager from AppSource](https://appsource.microsoft.com/en-us/product/office/WA104381209?tab=Overview).
2. Ensure a mail item is selected.
3. Open the LEAP Mail Manager. You should see a welcome screen with a Start button.
4. Click on the Start button. A login screen opens up in a dialog window.
5. Login with `officejs@leaptest.io / Pass2018`.
6. Notice, the dialog stays open on a grey background.
7. Attach the Microsoft Edge DevTools Preview debugger to the dialog.
8. Scroll all the way to the bottom, you should see a stacktrace with the error: `AgaveSink could not find a suitable sink to use`.

Now, to be fair, I am on the Office Insider channel, so my clients would not see this. However, the whole point of my being on the Insider channel is so that I can find issues like this before my customers do.

Please advise.

Updated list of versions:

> Platform: PC Desktop
> Host: Outlook
> Office version number: Outlook 2016 for Windows Version 1909 (Build 12026.20000 Click-to-Run), Office Insider
> Operating System: Windows 10 Version 1903 (OS Build 18362.295)

**Comment by jargil** [comment]
@TheSamsterZA - thanks for the additional repro steps. Taking a look.

**Comment by jargil** [comment]
@TheSamsterZA - I can confirm I can reproduce with the "LEAP Mail" app. Will debug through and update once I have something.

**Comment by jargil** [comment]
@TheSamsterZA - so there's definitely a bug that I'll address in Office's code. While I finalize the fix, and the fix gets deployed, let me explain what's going on and see if you can work-around.

You stated (https://github.com/OfficeDev/office-js/issues/633#issuecomment-515245302) that after you open the dialog, you navigate to a different page "dialog.html on the same domain", and then redirect to the main page on the taskpane. I assumed that everything was hosted on the same domain. That works fine. However, the problem is if dialog.html and the main page on the taskpane are on different domains. Which is what is happening with the "LEAP Mail" add-in.

You should be able to work-around until the fix is deployed by having a double redirection flow: popup the dialog with the taskpane domain, navigate to your auth domain (auth.leap.services) and then do the final redirection where you call messageParent in the original taskpane domain. Basically you want the first url with which you open the dialog to be the one that will be messaging the parent. 

I'll update once I have a build number with the fix.

@YannickRe FYI.

**Comment by jargil** [comment]
@TheSamsterZA - just a quick update regarding which build will have the fix for you to try: 16.0.12105 or newer. Should begin deploying to Office Insiders sometime next week, unfortunately I just missed the window for this week's update. Thanks again for all your input and feedback.

**Comment by jargil** [comment]
@TheSamsterZA - build 16.0.12105 is now deploying to Office Insiders. Please let me know if you can confirm the fix works for you. Thanks!

**Comment by TheSamsterZA** [comment]
@jargil I can confirm that the issue I was seeing before, `AgaveSink could not find a suitable sink to use`, has been fixed! I can log in via the OfficeJS dialog on the LEAP Mail Manager successfully, and I receive my auth token. I also did a quick test with my simple repro add-in, and all is well there, too.

Thank you for sticking with this and finding the solution. 😃
